design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,150 @@
+  when a built-in operation is overridden, you end up losing all existing functionality,
+  like uploading/downloading the files from the cloud as well.
+
+The goal is to provide an interface that allows users to extend only the specific aspect of the operation that they want,
+like the pre-update or post-update actions, leaving the rest of the behavior unchanged.
+
+## Solution proposal
+
+- The agent supports plugins for log and config management.
+- The supported types for the respective operation is gathered from all these plugins.
+- The agent maintains a mapping of the types and their corresponding plugins.
+- When an an operation is received for a type, it is delegated to the corresponding plugin.
+
+### Log Management
+
+- `tedge-agent` to add support for multiple log management plugins defined under `/etc/tedge/log-plugins`.
+- Migrate the existing file-based log functionality into a dedicated `file` plugin which is available by default.
+- To retrieve logs from other sources like systemd journal or docker, add respective plugins for these sources.
+- Each plugin must support the following sub-commands:
+  - `list`: list all the log types that it supports.
+    Used to detect if this is a valid plugin if it exits with exit code 0.
+    The output is used only when auto-discovery for log types is turned ON for this plugin.
+  - `get <type> [--from <timestamp>] [--to <timestamp>] [--filter <filter-text>]`:
+    Used to fetch the log for the given type within the provided log range
+- The plugin directory would also have a `conf.d` directory where each plugin can store their configuration in a toml file.
+  These config files are used to turn type discovery on/off and to define inclusion/exclusion list when type discovery is turned on:
+  ```docker.toml
+  [docker]
+  auto_discover=true
+
+  [docker.include]
+  container_name_regex = "nginx"
+
+  [docker.include]
+  image_name_regex = ""
+  
+  [docker.exclude]
+  container_name_regex = "kube-*"
+  ``` 
+- The main config file for a plugin must be named after the plugin itself (e,g: `docker.sh` -> `docker.toml`).
+  Key configs like `auto_discover` can only be defined in this main file.
+  The inclusion/exclusion list can be defined in extension files as well,
+  but the table names with the type prefix (e.g: `[docker.include]` and `[docker.exclude]`) must be used in those as well.
+- When new software is installed, their corresponding `include/exclude` entries can be appended to the main plugin config itself,
+  or created in an independent extension file.
+
+## Config management
+
+### Phase 1
+
+- Similar to log management, config plugins are defined under `/etc/tedge/config-plugins`.
+- Each plugin corresponds to their corresponding config types, unlike the log plugins where each plugin supports multiple types.
+- A config plugin needs to support the following sub-commands:
+  - `info`: Used to detect if this is a valid plugin if it exits with exit code 0.
```

Comment:
It might be best to re-use existing concepts, and use the `list` sub command which provides by the list of configurations which are associated with the given config type plugin, as we need to get the list of types anyway at some point in time...though this way, the tedge-configuration-plugin configuration type, when requested, would return a dynamic content rather then the static file contents of the tedge-configuration-plugin.toml as it would need to query each plugin to see what configuration types are supported...which then opens up the question, when is the list checked?

Reply:
It was decided to keep the config plugin interfaces similar to that of log plugins in the [spec](7cdd217175e0ec0be02ad7bf30d09a2fa3c8f127)